### ISSUE TITLE
Comparison with Nx.js

### ISSUE THREAD

**Issue by alexkreidler** [microsoft/rushjs.io-website#76]
Nx is I think one of the projects with the most similar scope as Rush. One of their big selling points is ["monorepos at scale"](https://nx.dev/latest/react/guides/monorepo-nx-enterprise), and thus have features like
- cross-project command running (including in parallel)
- nx affected:build (uses project caching, a deps graph, and git status to only rebuild affected)

These are also core features of Rush. In fact, since they're so similar in these aspects, I'd be very interested to see a comparison of their performance. I suspect Rush by default is faster on installs due to PNPM, but I'm very curious about the change detection and task running performance. Also, I'm curious if Rush is any faster even using NPM or Yarn.

Nx also takes on somewhat of a larger scope, also trying to manage webpack, jest, and typescript configs with their own presets under the `@nrwl` npm org. Personally, my experience as a relatively new Nx user is that I've spent unordinate amounts of time going to the [nx monorepo](https://github.com/nrwl/nx) and trying to read through webpack, babel, and jest configs to figure out what's going on. One other concrete difference is that [Nx promotes a pattern of using one `package.json` for the entire monorepo](https://github.com/nrwl/nx/issues/1777), meaning each dependency _must_ be at the same version for all packages.

After reading the Rush docs, I'm inclined to switch over, because Rush handles all the phantom dependency issues, is unopinionated on project-level build configs, and also, Rush is built with Rush! Also, looking at the contributors to both projects, I see Rush's top five contributors are on a similar level, while Nx's top two far outstrip the others. This gives me some confidence that Rush is truly maintained, and hearing about Rush in production does the same.

I guess the real request here is
- Add a qualitative comparison with Nx to the docs. This could be in the same file as #60 
- (Possibly) add a performance comparison to Nx for running tasks across monorepos, or testing affected libraries. This would require two large-enough monorepos to be created for Nx and Rush, which would probably be considerable work, so this is less important.

**Comment by jeffbcross** [comment]
Your contributor comparison isn't accurate. If you look at the past year of both projects, Nx and Rush both have two contributors who are at least 2x the number of commits of any other contributor, with many other significant contributors with substantial contributions (40 other contributors for rushstack, 66 other contributors for Nx). 

![github com_microsoft_rushstack_graphs_contributors_from=2019-11-17 to=2020-11-16 type=c](https://user-images.githubusercontent.com/463703/99835228-89c16480-2b21-11eb-8819-11b13c1c9dee.png)
![github com_nrwl_nx_graphs_contributors_from=2019-11-17 to=2020-11-16 type=c](https://user-images.githubusercontent.com/463703/99835242-8e861880-2b21-11eb-9043-1b79f14369e7.png)

Contributors are a hard thing to compare apples to apples across teams, since there are so many variables in how the teams work, what things are developed inside the main project repo or outside the repo, how much the teams squash commits vs smaller logical commits.

I work at Nrwl (company behind Nx), and we have nothing but respect for the Rush team. I'm sure both Nx and Rush are used in many large workspaces for production applications. The choice between the two is probably a matter of preference (like Nx's prescriptiveness of a Single Version Policy for example).

**Comment by tmtron** [comment]
Well, for us the most important limitation of nx is the single package.json restriction. What I mean exactly is this:
* Single Version Policy is great and we want to keep it (i.e. when a dependency is used in the monorepo, we want to have the same version in all applications/libraries that use it)
* But every application/library **must** have a way to specify only the dependencies (libs/packages) that it really needs (this avoids many problems)

AFAIU, in rushjs we can have a package.json file per project and use `rush check` to ensure that we have one consistent version per dependency. This should be mentioned in a comparison list.

**Comment by jeffbcross** [comment]
> Well, for us the most important limitation of nx is the single package.json restriction.
It's a not a restriction of Nx, it's just the default. You can have package.json within each project if you'd like. Or you could include multiple versions of a dependency in the root package.json.

**Comment by octogonz** [comment]
> Contributors are a hard thing to compare apples to apples across teams, since there are so many variables in how the teams work, what things are developed inside the main project repo or outside the repo, how much the teams squash commits vs smaller logical commits.

I agree 100% with this. Even if GitHub stats were intelligent (which they are not), commit diffs simply aren't representative of a person's quality of work or even volume of work.  For example, the person who ran Prettier to bulk-edit thousands of files might appear like a hero in the stats, whereas the actual hero may have relatively few commits. I remember a person whose GitHub profile showed tons of activity that, on closer inspection, includes a lot of automated package.json version bumps.  Turns out he configured his publishing bot to run using a personal token. 😁

Better to look at things like how long has the project been around?  Who are the people behind it? Are they contributing professionally or as a weekend hobby?  How mature is their process for design discussions and code reviews?  Nx and Rush are probably fairly comparable in all these areas.

> we have nothing but respect for the Rush team. I'm sure both Nx and Rush are used in many large workspaces for production applications. The choice between the two is probably a matter of preference (like Nx's prescriptiveness of a Single Version Policy for example).

I prefer to avoid writing competitive reports showing how we stack up against the "competition."  Open source has a different character from a commercial product, even when we get paid to work on it.  You'll find the mentality is less competitive, more about finding a particular niche to specialize in, and about solving problems in a way that aligns with a particular set of requirements or philosophy.

It's certainly useful to have comparisons that help people decide what to use. But it may be better to leave that to unbiased parties.  Rush could certainly share links and general info about where it fits in the space, though.  (Although... if we mention Nx, would it be unfair not to mention 5 other similar projects? And since Microsoft supports the Rush projects... could people interpret these statements as company endorsements? Better be careful how we phrase it and triple-check every assertion. Maybe life is simpler if we just focus on making Rush great, and let someone else do those writeups. 🙂)

**Comment by alexkreidler** [comment]
> It's certainly useful to have comparisons that help people decide what to use. 

+1. That was my primary reason for creating this issue, to provide a starting point for comparisons based on my own research, not to criticize either project or make extensive comparisons from subjective metrics (e.g. Github contributor data).

I understand the desire to avoid any presumption of endorsement. However, in my experience, comparative writing in OSS docs can be one of the most persuasive technical factors influencing my decision to use software. All the [best](https://www.terraform.io/intro/vs/chef-puppet.html) [examples](https://medium.com/@dan_abramov/you-might-not-need-redux-be46360cf367) of this really focus on describing their niche and features while comparing against projects in a similar space in a non-competitive way. E.g.:
> finding a particular niche to specialize in, and about solving problems in a way that aligns with a particular set of requirements or philosophy.

I do understand the hesitation to go down the road of writing lots of informative "articles" that are slightly biased towards your solution that many tech startups seem to do. However there is a way to write good technical comparison docs that are clear, unbiased, and manifestly helpful.

I'd much rather read a comparison with healthy skepticism than have to take the time to try both (or 3 or more) pieces of software to gain the appropriate insights. I think I'm on the extreme end where I dig deep into projects before diving in, so it is sometimes frustrating to run into a lack of clarity in the docs and end up using one solution or another for a while until running into its limits.

Anyways, best of luck with everything, I hope this hasn't been too much of a distraction from all the great work both projects are doing!

**Comment by octogonz** [comment]
These are good points.

**Comment by elliottsj** [comment]
I wrote a short post in November comparing some aspects of Rush and Nx, primarily npm dependency management: https://elliott.dev/posts/rush-vs-nx

It's admittedly slightly tilted towards Rush, however I tried to remain fairly objective and mention Nx's strengths as well.

**Comment by jonathan-irvin** [comment]
> I wrote a short post in November comparing some aspects of Rush and Nx, primarily npm dependency management: https://elliott.dev/posts/rush-vs-nx
> 
> 
> 
> It's admittedly slightly tilted towards Rush, however I tried to remain fairly objective and mention Nx's strengths as well.

One year later, your article helped a bunch!
